<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="primitive_chargepal_with_fix_plug">

  <!-- Macros -->
  <xacro:include filename="$(find chargepal_description)/urdf/utils/chargepal_common.urdf.xacro" />

  <!-- constants -->
  <xacro:property name="M_PI" value="3.14159"/>

  <!-- Resolve prefix variable (useful for multi robot) -->
  <xacro:include filename="$(find chargepal_description)/urdf/utils/prefix.urdf.xacro" />


  <!-- ############### PLATFORM ################ -->
  <xacro:include filename="$(find chargepal_description)/urdf/mobile_platform/primitive_mobile_platform.urdf.xacro"/>

  <!-- platform arm joint -->
  <joint name="${prefix}platform_ur10e_joint" type="fixed">
    <parent link="${prefix}tower_link" />
    <child link="${prefix}base_link" />
    <origin xyz="0.0 0.0 ${tower_height/2}" rpy="0.0 0.0 ${0.5*M_PI}" />
  </joint>


  <!-- ############### UR10e ARM ################ -->
  <xacro:include filename="$(find ur_description)/urdf/inc/ur10e_macro.xacro"/>
  <!-- Arguments to load UR10e arm macro -->
  <xacro:arg name="joint_limit_params" default="$(find chargepal_description)/config/ur10e/joint_limits.yaml"/>
  <xacro:arg name="physical_params" default="$(find ur_description)/config/ur10e/physical_parameters.yaml"/>
  <xacro:arg name="kinematics_params" default="$(find ur_description)/config/ur10e/default_kinematics.yaml"/>
  <xacro:arg name="visual_params" default="$(find ur_description)/config/ur10e/visual_parameters.yaml"/>
  <!--
    legal values:
      - hardware_interface/PositionJointInterface
      - hardware_interface/VelocityJointInterface
      - hardware_interface/EffortJointInterface

    NOTE: this value must correspond to the controller configured in the
          controller .yaml files in the 'config' directory.
  -->
  <xacro:arg name="transmission_hw_interface" default="hardware_interface/VelocityJointInterface"/>  
  <xacro:arg name="safety_limits" default="false"/>
  <xacro:arg name="safety_pos_margin" default="0.15"/>
  <xacro:arg name="safety_k_position" default="20"/>

  <xacro:ur10e_robot
    prefix="${prefix}"
    joint_limits_parameters_file="$(arg joint_limit_params)"
    kinematics_parameters_file="$(arg kinematics_params)"
    physical_parameters_file="$(arg physical_params)"
    visual_parameters_file="$(arg visual_params)"
    transmission_hw_interface="$(arg transmission_hw_interface)"
    safety_limits="$(arg safety_limits)"
    safety_pos_margin="$(arg safety_pos_margin)"
    safety_k_position="$(arg safety_k_position)"
  />

  <!-- Configure self collision properties per link -->
  <gazebo reference="${prefix}shoulder_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}upper_arm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}forearm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_3_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_2_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}ee_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  

  <!-- ############### Bota Systems SensONE FT SENSOR-->
  <xacro:include filename="$(find rokubimini_description)/urdf/BFT_SENS_ECAT_M8.urdf.xacro" />

  <xacro:BFT_SENS_ECAT_M8 prefix="${prefix}ft_sensor" />

  <joint name="ur_to_sensor" type="fixed">
    <parent link="${prefix}tool0" />
    <child link="${prefix}ft_sensor_mounting" />
    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
  </joint>

  <!-- ############### FIX PLUG ################ -->
  <xacro:include filename="$(find chargepal_description)/urdf/tools/primitive_fix_plug.urdf.xacro" />

  <xacro:fix_plug parent="${prefix}ft_sensor_wrench" prefix="${prefix}" />

  <!-- Add three (for each axis) additional virtual frames for reward calculation -->
  <xacro:chargepal_virtual_joint
    joint_name="${prefix}virt_tool_frame_x_joint" 
    type="fixed"
    parent="${prefix}plug_reference_frame"
    child="${prefix}virt_tool_frame_x"
    joint_axis_xyz="0 0 0"
    joint_origin_xyz="1 0 0"
    joint_origin_rpy="0 0 0"
    joint_lower_limit="0"
    joint_upper_limit="0"
  />
  <xacro:chargepal_virtual_link link_name="${prefix}virt_tool_frame_x" />

  <xacro:chargepal_virtual_joint
    joint_name="${prefix}virt_tool_frame_y_joint" 
    type="fixed"
    parent="${prefix}plug_reference_frame"
    child="${prefix}virt_tool_frame_y"
    joint_axis_xyz="0 0 0"
    joint_origin_xyz="0 1 0"
    joint_origin_rpy="0 0 0"
    joint_lower_limit="0"
    joint_upper_limit="0"
  />
  <xacro:chargepal_virtual_link link_name="${prefix}virt_tool_frame_y" />

  <xacro:chargepal_virtual_joint
    joint_name="${prefix}virt_tool_frame_z_joint" 
    type="fixed"
    parent="${prefix}plug_reference_frame"
    child="${prefix}virt_tool_frame_z"
    joint_axis_xyz="0 0 0"
    joint_origin_xyz="0 0 1"
    joint_origin_rpy="0 0 0"
    joint_lower_limit="0"
    joint_upper_limit="0"
  />
  <xacro:chargepal_virtual_link link_name="${prefix}virt_tool_frame_z" />


  <!-- ################ GAZEBO ################ -->
  <!-- Common -->
  <xacro:include filename="$(find chargepal_description)/urdf/utils/common.gazebo.xacro" />
  <xacro:controller_plugin_gazebo robot_namespace=""/>

</robot>